awxkit/awxkit/cli/custom.py

```diff
@@ -196,6 +197,116 @@ def perform(self, slug):
         return self.page.get()
 
 
+class RoleMixin(object):
+
+    has_roles = [
+        ['organizations', 'organization'],
+        ['projects', 'project'],
+        ['inventories', 'inventory'],
+        ['inventory_scripts', 'inventory_script'],
+        ['teams', 'team'],
+        ['credentials', 'credential'],
+        ['job_templates', 'job_template'],
+        ['workflow_job_templates', 'workflow_job_templates'],
+    ]
+    roles = {}  # this is calculated once
+
+    def add_arguments(self, parser):
+
+        if not RoleMixin.roles:
+            for resource, flag in self.has_roles:
+                options = self.page.__class__(
+                    self.page.endpoint.replace(self.resource, resource),
+                    self.page.connection
+                ).options()
+                RoleMixin.roles[flag] = [
+                    role.replace('_role', '')
+                    for role in options.json['object_roles']
+                ]
+
+        possible_roles = set()
+        for v in RoleMixin.roles.values():
+            possible_roles.update(v)
+
+        resource_group = parser.choices[self.action].add_mutually_exclusive_group(
+            required=True
+        )
+        parser.choices[self.action].add_argument(
+            'id', type=int,
+            help='The ID of the {} to {} access to/from'.format(
+                self.resource, self.action
+            )
+        )
+        for _type in RoleMixin.roles.keys():
+            if _type == 'team' and self.resource == 'team':
+                # don't add a team to a team
+                continue
+            resource_group.add_argument(
+                '--{}'.format(_type), type=int, metavar='ID',
+                help='The ID of the target {}'.format(_type),
+            )
+        parser.choices[self.action].add_argument(
+            '--type', type=str, choices=possible_roles, required=True,
+            help='The name of the role to {}'.format(self.action)
+        )
+
+    def perform(self, **kwargs):
+        for resource, flag in self.has_roles:
+            if flag in kwargs:
+                _type = kwargs['type']
+                if _type not in RoleMixin.roles[flag]:
```

Comment:
I may have to disabuse myself of prior assumptions here. When we say that this CLI is OPTIONS based, one thing that means is that an OPTIONS request has to complete to render the `--help` text. When a `awx credential create` or an `awx user grant` command is ran, it was not obvious to me that an OPTIONS request would precede the POST request to create the resource or association.

This points me in a new direction, since it seems to be assumed that `add_arguments` is always called before this method within a single CLI invocation. In other words, this way this is structured requires that no request can be done without doing an OPTIONS request prior to it in order to validate (and sometimes transform?) the inputs.

I realize this is a very basic level question, but is that all consistent with how this is designed?

Reply:
That's *generally* the workflow for most CLI commands - before actually *doing* the underlying request (making/editing/fetching) something, we do an HTTP OPTIONS to verify that the inputs are *actually valid in the first place*.